packages/fxa-auth-server/lib/payments/paypal-processor.ts

```diff
@@ -0,0 +1,331 @@
+
+  private attemptsToday(transactions: TransactionSearchResult[]) {
+    const today = new Date();
+    return transactions.filter((t) => sameDay(today, new Date(t.timestamp)))
+      .length;
+  }
+
+  /**
+   * Attempt to process an invoice that is at least 24 hours after its creation.
+   *
+   * This attempt consists of:
+   *   1. Verifying the customer was expanded on the invoice.
+   *   2. Ensuring we don't process invoices for deleted customers, their invoice will
+   *      be marked as uncollectible.
+   *   3. Search all transactions run for this invoice and:
+   *     a. Ensure the transaction count matches the invoice retry attempts.
+   *     b. If a transaction completed, mark invoice as paid, return.
+   *     c. If a transaction is pending/in-progress:
+   *       - If we're in grace period, skip invoice and wait for completion, return.
+   *   4. If we're past grace period, cancelInvoiceSubscription, return.
+   *   5. If we have no billing agreement, email user to inform them.
+   *   6. If retries for today remain:
+   *       a. Attempt payment on invoice.
+   *   7. Otherwise, return.
+   *
+   * @param invoice
+   */
+  private async attemptInvoiceProcessing(invoice: Stripe.Invoice) {
+    const customer = invoice.customer;
+    // 1
+    if (!customer || typeof customer === 'string') {
+      this.log.error('customerNotLoaded', { customer });
+      throw error.internalValidationError('customerNotLoad', {
+        customer,
+        invoiceId: invoice.id,
+      });
+    }
+
+    // 2, Void the invoice for deleted customers
+    if (customer.deleted) {
+      await this.stripeHelper.markUncollectible(invoice);
+      this.log.info('customerDeletedVoid', { customerId: customer.id });
+      return;
+    }
+
+    // 3, search transactions
+    const transactions = await this.paypalHelper.searchTransactions({
+      startDate: new Date(invoice.created * 1000),
+      invoice: invoice.id,
+    });
+    // 3a
+    await this.ensureAccurateAttemptCount(invoice, transactions);
+
+    // 3b
+    if (await this.handlePaidTransaction(invoice, transactions)) {
+      return;
+    }
+
+    // 3c
+    if (await this.handlePendingTransaction(invoice, transactions)) {
+      return;
+    }
+
+    // 4
+    const inGracePeriod = this.inGracePeriod(invoice);
+    if (!inGracePeriod) {
+      return this.cancelInvoiceSubscription(invoice);
+    }
+
+    // 5
+    const billingAgreementId = this.stripeHelper.getCustomerPaypalAgreement(
+      customer
+    );
+    if (!billingAgreementId) {
+      // TODO: Send email to user that they must go to sub management and re-auth PayPal
+      //       Don't send email if payment attempts today > 1.
+      return;
+    }
+
+    // 6
+    if (this.attemptsToday(transactions) < this.maxRetryAttempts) {
+      await this.makePaymentAttempt(invoice);
+    }
+    return;
+  }
+
+  public async processInvoices() {
+    // Generate a time 6 hour prior.
+    const sixHoursAgo = hoursBeforeInSeconds(6);
```

Comment:
Any reason we don't want the #hour value passed in?

Reply:
Not really, this could be a default passed in.